sklearn/metrics/tests/test_pairwise_distances_reduction.py

```diff
@@ -96,24 +193,351 @@ def assert_radius_neighborhood_results_equality(ref_dist, dist, ref_indices, ind
+    rounding errors on the distance computations.
+
+    Input arrays must be sorted w.r.t distances.
+    """
+    is_sorted = lambda a: np.all(a[:-1] <= a[1:])
+
+    n_significant_digits = -(int(floor(log10(abs(rtol)))) + 1)
+
+    assert (
+        len(ref_dist) == len(dist) == len(ref_indices) == len(indices)
+    ), "Arrays of results have various lengths."
+
+    n_queries = len(ref_dist)
+
+    # Asserting equality of results one vector at a time
+    for query_idx in range(n_queries):
+
+        ref_dist_row = ref_dist[query_idx]
+        dist_row = dist[query_idx]
+
+        assert is_sorted(
+            ref_dist_row
+        ), f"Reference distances aren't sorted on row {query_idx}"
+        assert is_sorted(dist_row), f"Distances aren't sorted on row {query_idx}"
+
+        # Vectors' lengths might be different due to small
+        # numerical differences of distance w.r.t the `radius` threshold.
+        largest_row = ref_dist_row if len(ref_dist_row) > len(dist_row) else dist_row
+
+        # For the longest distances vector, we check that last extra elements
+        # that aren't present in the other vector are all in: [radius ± atol]
+        atol = 10 ** (-n_significant_digits)
+        min_length = min(len(ref_dist_row), len(dist_row))
+        last_extra_elements = largest_row[min_length:]
+        if last_extra_elements.size > 0:
+            assert np.all(radius - atol <= last_extra_elements <= radius + atol), (
+                f"The last extra elements ({last_extra_elements}) aren't in [radius ±"
+                f" atol]=[{radius} ± {atol}]"
+            )
+
+        # We truncate the neighbors results list on the smallest length to
+        # be able to compare them, ignoring the elements checked above.
+        ref_dist_row = ref_dist_row[:min_length]
+        dist_row = dist_row[:min_length]
+        print(type(ref_dist_row))
+        print(type(dist_row))
+
+        assert_allclose(ref_dist_row, dist_row, rtol=rtol)
+
+        ref_indices_row = ref_indices[query_idx]
+        indices_row = indices[query_idx]
+
+        # Grouping indices by distances using sets on
+        # a rounded distances up to a given number of decimals
+        reference_neighbors_groups = defaultdict(set)
+        effective_neighbors_groups = defaultdict(set)
+
+        for neighbor_rank in range(min_length):
+            rounded_dist = relative_rounding(
+                ref_dist_row[neighbor_rank],
+                n_significant_digits=n_significant_digits,
+            )
+            reference_neighbors_groups[rounded_dist].add(ref_indices_row[neighbor_rank])
+            effective_neighbors_groups[rounded_dist].add(indices_row[neighbor_rank])
+
+        # Asserting equality of groups (sets) for each distance
+        msg = (
+            f"Neighbors indices for query {query_idx} are not matching "
+            f"when rounding distances at n_significant_digits={n_significant_digits}"
```

Comment:
Since `n_significant_digits` is no longer part of the public API of this helper, let's update the message accordingly:

```suggestion
            f"when rounding distances at {n_significant_digits} significant digits "
            f"derived from rtol={rtol:.1e}"
```